File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        HashSet<Character>set = new HashSet<>();
        int left =0;
        int longestSubStringLength =0;
        for(int right=0;right<n;right++){
            while(set.contains(s.charAt(right))){
                set.remove(s.charAt(left));
                left++;
            }
            set.add(s.charAt(right));
            longestSubStringLength = Math.max(longestSubStringLength, right-left+1);
        }
        return longestSubStringLength;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = int(s.length()), res = 0;
        unordered_map<char, int> charToNextIndex;

        for (int j = 0, i = 0; j < n; j++) {
            if (charToNextIndex[s[j]] > 0) {
                i = max(charToNextIndex[s[j]], i);
            }
            res = max(res, j - i + 1);
            charToNextIndex[s[j]] = j + 1;
        }
        return res;
    }
};